File: 882-peak-index-in-a-mountain-array/peak-index-in-a-mountain-array.cpp
class Solution {
public:
    int peakIndexInMountainArray(vector<int>& arr) {
        int i = 0, j = arr.size()-1;
        while (i <= j){
            int mid = midpoint(i, j);

            if (mid > 0 && arr[mid] < arr[mid-1]) j = mid - 1;
            else if (mid < arr.size()-1 && arr[mid] < arr[mid+1]) i = mid + 1;
            else return mid;
        }

        return -1;
    }
};

File: 882-peak-index-in-a-mountain-array/peak-index-in-a-mountain-array.go
func peakIndexInMountainArray(nums []int) int {
    i, j := 1, len(nums)-2
    for i <= j {
        mid := i + (j - i)/2

        if nums[mid] > nums[mid-1] && nums[mid] > nums[mid+1] {
            return mid
        } else if nums[mid] > nums[mid-1] && nums[mid] < nums[mid+1] {
            i = mid + 1
        } else {
            j = mid -1
        }
    }

    return -1
}